src/analysis/souffle/taint.dl

```diff
@@ -34,48 +34,82 @@
+// independently spread through the graph from that point according to their different behavior
+// patterns.
+.type ClaimedTag <: symbol
 
 // Is this symbol an access path to some data?
 // (Useful for scoping when negation is involved.)
 .decl isAccessPath(accessPath: AccessPath)
 
-// Is this a Tag?
+// The set of all privacy tags.
 // (Useful for scoping when negation is involved.)
-.decl isTag(tag: Tag)
+.decl allPrivacyTags(privacyTag: PrivacyTag)
+
+// The set of all integrity tags.
+// (Useful for scoping when negation is involved.)
+.decl allIntegrityTags(integrityTag: IntegrityTag)
+
+// The set of all claimed tags.
+// (Useful for scoping when negation is involved.)
+.decl allClaimedTags(claimedTag: ClaimedTag)
 
 // A data flow edge.
 .decl edge(src: AccessPath, tgt: AccessPath)
 
 // A direct or transitive data flow path.
 .decl path(src: AccessPath, tgt: AccessPath)
 
-// Declares that `accessPath` may have be tainted with `tag`.
-// If false, `accessPath` *definitely* does not `tag` taint at runtime.
-// If true, `accessPath` may have the `tag` taint at runtime.
-.decl mayHaveTag(accessPath: AccessPath, tag: Tag)
+// Declares that `accessPath` may have be tainted with `privacyTag`.
+// If false, `accessPath` *definitely* does not `privacyTag` taint at runtime.
+// If true, `accessPath` may have the `privacyTag` taint at runtime.
+.decl mayHavePrivacyTag(accessPath: AccessPath, privacyTag: PrivacyTag)
 
-// There is a claim that `accessPath` definitely has `tag` taint.
-.decl claimHasTag(accessPath: AccessPath, tag: Tag)
+// Declares that `accessPath` must be tagged with `integrityTag`.
+// If false, `accessPath` may or may not have the integrity property indicated by tag at runtime.
+// If true, `accessPath` *definitely* has the integrity property indicated by tag at runtime.
+.decl mustHaveIntegrityTag(accessPath: AccessPath, integrityTag: IntegrityTag)
+
+// There is a claim that `accessPath` definitely has `tag`. This introduces *both* a privacy and
+// integrity tag.
+.decl claimHasTag(accessPath: AccessPath, claimedTag: ClaimedTag)
 
 // Some inference rules
 
 // Symbols used in edges are access paths
 isAccessPath(x) :- edge(x, _).
 isAccessPath(y) :- edge(_, y).
 
-// Symbols used in hasTag or mayHaveTag are tags
-isTag(tag) :- claimHasTag(_, tag).
-isTag(tag) :- mayHaveTag(_, tag).
+allPrivacyTags(as(claimedTag, PrivacyTag)) :- claimHasTag(_, claimedTag).
+allPrivacyTags(privacyTag) :- mayHavePrivacyTag(_, privacyTag).
+
+allIntegrityTags(as(claimedTag, IntegrityTag)) :- claimHasTag(_, claimedTag).
+allIntegrityTags(integrityTag) :- mustHaveIntegrityTag(_, integrityTag).
 
 // Transitive paths
 path(from, to) :- edge(from, to).
 path(from, to) :- edge(from, intermediate), path(intermediate, to).
 
-mayHaveTag(tgt, tag) :- claimHasTag(tgt, tag).
-mayHaveTag(tgt, tag) :- edge(src, tgt), mayHaveTag(src, tag).
+mayHavePrivacyTag(tgt, as(claimedTag, PrivacyTag)) :- claimHasTag(tgt, claimedTag).
```

Comment:
what does `as(claimedTag, PrivacyTag)` do?

Reply:
`as` is a cast from one base type with a particular backing primitive type into another base type with the same backing primitive type. In this case, I'm saying "take the value of `claimedTag` and treat it as a `PrivacyTag` so long as it also is backed by `symbol`."